src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1976,12 +2039,8 @@ internal Exception Break(Exception reason)
                 DataSource.Clear();
             }
 
-            LogMessages.BreakingConnection(ConnectionLogger, Id, reason);
-
-            // Note that we may be reading and writing from the same connector concurrently, so safely set
-            // the original reason for the break before actually closing the socket etc.
-            Interlocked.CompareExchange(ref _breakReason, reason, null);
-            State = ConnectorState.Broken;
+            // Make sure there is no concurrent cancellation in process
+            lock (CancelLock) { }
```

Comment:
Any specific reason to do this, rather than just have CancelLock taken throughout the method as before?

Doesn't this allow Cancel to be called right after this, and so race with FullCleanup?

Reply:
> Any specific reason to do this, rather than just have CancelLock taken throughout the method as before?

We can, though there is no point. Any cancellation request after `SingleUseLock` exited here should exit immediately.

> Doesn't this allow Cancel to be called right after this, and so race with FullCleanup?

Nope. The first thing Cancel does is to take `SingleUseLock` (which it should) and check the state. At this point it's going to be `Broken`, so it's just going to exit.

Reply:
OK.

Your logic seems correct, but it's also a bit more complex to see when actually looking at the code (after all, I needed you to explain it :sweat_smile:). I'm saying we should try to be as simple as possible with this concurrency stuff - and I think just taking the lock over the entire operation makes it simpler for the external person reading the code.

But up to you.